File: front-end/src/app/domain/models/game/GameSession.ts
import {GameAction} from "./action/GameAction";
import {Player} from "./Player";
import {GameActionUtil} from "./action/GameActionUtil";
import {GameBoard} from "./board/GameBoard";
import {PropertyDevelopmentStageTracker} from "./board/location/locationImpl/PropertyDevelopmentStageTracker";

export enum GameSessionState {
  JOINABLE = "JOINABLE",
  STARTED = "STARTED",
  COMPLETED = "COMPLETED"
}

export class GameSession {
  private _id: number;

  constructor(
    id: number,
    maxPlayers: number,
    startsAt: Date,
    maxTimeTurn: number,
    endsAt: Date,
    players: Player[],
    actions: GameAction[],
    creator: string,
    state: GameSessionState,
    gameBoard: GameBoard,
    currentTurn: number,
    propertyDevelopments: PropertyDevelopmentStageTracker[]
  ) {
    this._id = id;
    this._maxPlayers = maxPlayers;
    this._startsAt = startsAt;
    this._maxTimeTurn = maxTimeTurn;
    this._endsAt = endsAt;
    this._players = players;
    this._actions = actions;
    this._creator = creator;
    this._state = state;
    this._gameBoard = gameBoard;
    this._currentTurn = currentTurn;
    this._propertyDevelopments = propertyDevelopments;
  }

  private _startsAt: Date;

  /**
   * Maximum amount of players. Min=2. Max=4.
   */
  private _maxPlayers: number;

  private _endsAt: Date;

  public get maxPlayers(): number {
    return this._maxPlayers;
  }

  /**
   * Time between turns in minutes.
   */
  private _maxTimeTurn: number;

  public get maxTimeTurn(): number {
    return this._maxTimeTurn;
  }

  private _gameBoard: GameBoard;

  public get gameBoard(): GameBoard {
    return this._gameBoard;
  }

  private _currentTurn: number;

  public get currentTurn(): number {
    return this._currentTurn;
  }

  public set currentTurn(value: number) {
    this._currentTurn = value;
  }

  private _players: Player[];

  public get players(): Player[] {
    return this._players;
  }

  public set players(value: Player[]) {
    this._players = value;
  }

  private _actions: GameAction[];

  public get actions(): GameAction[] {
    return this._actions;
  }

  public set actions(value: GameAction[]) {
    this._actions = value;
  }

  private _creator: string;

  public get creator(): string {
    return this._creator;
  }

  private _propertyDevelopments: PropertyDevelopmentStageTracker[];

  public get propertyDevelopments(): PropertyDevelopmentStageTracker[] {
    return this._propertyDevelopments;
  }

  public get id(): number {
    return this._id;
  }

  public get endsAt(): Date {
    return this._endsAt;
  }

  private _state: GameSessionState;

  public get state(): GameSessionState {
    return this._state;
  }

  public get startsAt(): Date {
    return this._startsAt;
  }

  public static trueCopy(session: GameSession): GameSession {
    return new GameSession(
      session.id,
      session.maxPlayers,
      new Date(session.startsAt),
      session.maxTimeTurn,
      new Date(session.endsAt),
      session.players.map(Player.trueCopy),
      session.actions.map(GameActionUtil.trueCopy),
      session.creator,
      session.state,
      GameBoard.trueCopy(session.gameBoard),
      session.currentTurn,
      session.propertyDevelopments.map(PropertyDevelopmentStageTracker.trueCopy)
    );
  }

  public toJson(): object {
    return {
      "id": this.id,
      "maxPlayers": this.maxPlayers,
      "startsAt": this.startsAt.toJSON(),
      "maxTimeTurn": this.maxTimeTurn,
      "endsAt": this.endsAt.toJSON(),
      "players": this.players.map(item => item.toJson()),
      "actions": this.actions.map(item => item.toJson()),
      "state": this.state,
      "gameBoard": this.gameBoard.toJson(),
      "currentTurn": this.currentTurn,
      "propertyDevelopments": this.propertyDevelopments.map(item => item.toJson)
    }
  }
}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/GameSession.java
package nl.hva.backend.domain.models.game;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.action.GameAction;
import nl.hva.backend.domain.models.game.board.GameBoard;
import nl.hva.backend.domain.models.game.board.location.Location;
import nl.hva.backend.domain.models.game.board.location.locationImpl.Property;
import org.springframework.data.annotation.CreatedBy;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Entity
@EntityListeners(AuditingEntityListener.class)

@Getter
@Setter
@NoArgsConstructor

@JsonIdentityInfo(
        generator = ObjectIdGenerators.PropertyGenerator.class,
        property = "id",
        scope = GameSession.class
)
public class GameSession {
    public static final int MAX_GAME_PARTICIPANTS = 4;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * Maximum amount of players. Min=2. Max=4.
     */
    private Integer maxPlayers;

    private LocalDateTime startsAt;

    /**
     * Time between turns in minutes.
     */
    private Integer maxTimeTurn;

    private LocalDateTime endsAt;

    @ManyToOne(fetch = FetchType.EAGER)
    private GameBoard gameBoard;

    private int currentTurn = 1;

    /**
     * Tracks the development of properties present on the board.
     * <p>
     * Each property on the board is required and can only have 1 tracker.
     */
    @OneToMany(cascade = CascadeType.ALL)
    private List<Property.PropertyDevelopmentStageTracker> propertyDevelopments;

    @Enumerated(EnumType.STRING)
    private GameSessionState state;

    @OneToMany(mappedBy = "game")
    private List<Player> players;

    @OneToMany(mappedBy = "gameSession")
    private List<GameAction> actions;

    @CreatedBy
    private String creator;

    private LocalDateTime startedAt;

    public enum GameSessionState {
        JOINABLE,
        STARTED,
        COMPLETED
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Location location)) return false;
        return getId() == location.getId();
    }

    @Override
    public int hashCode() {
        return getId() == null ? 0 : Long.hashCode(getId());
    }

    public void add(GameAction action) {
        if (getActions() == null || !(getActions() instanceof ArrayList))
            setActions(new ArrayList<>(Objects.requireNonNullElse(getActions(), List.of())));
        if (action == null) return;

        getActions().add(action);
    }

    public void add(Property.PropertyDevelopmentStageTracker tracker) {
        if (getPropertyDevelopments() == null || !(getPropertyDevelopments() instanceof ArrayList))
            setPropertyDevelopments(new ArrayList<>(Objects.requireNonNullElse(getPropertyDevelopments(), List.of())));

        getPropertyDevelopments().add(tracker);
    }

    public void add(Player player) {
        if (getPlayers() == null || !(getPlayers() instanceof ArrayList))
            setPlayers(new ArrayList<>(Objects.requireNonNullElse(getPlayers(), List.of())));

        getPlayers().add(player);
    }

}
